### PR TITLE
Joint Force-Torque Systems Plugin

### PR DESCRIPTION
**Pull request by Devansh16**
# 🎉 New feature

## Summary
Systems Plugin Implementation for Joint Force-Torque Sensor.

Needs https://github.com/ignitionrobotics/ign-sensors/pull/144

## Checklist
- [ ] Signed all commits for DCO
- [ ] Added tests
- [ ] Added example and/or tutorial
- [ ] Updated documentation (as needed)
- [ ] Updated migration guide (as needed)
- [ ] `codecheck` passed (See [contributing](https://ignitionrobotics.org/docs/all/contributing#contributing-code))
- [ ] All tests passed (See [test coverage](https://ignitionrobotics.org/docs/all/contributing#test-coverage))
- [ ] While waiting for a review on your PR, please help review [another open pull request](https://github.com/pulls?q=is%3Aopen+is%3Apr+user%3Aignitionrobotics+repo%3Aosrf%2Fsdformat+archived%3Afalse+) to support the maintainers

**Note to maintainers**: Remember to use **Squash-Merge**

### CODE REVIEW

File: src/systems/CMakeLists.txt [hunk 1]
```diff
@@ -101,6 +101,7 @@ add_subdirectory(diff_drive)
 add_subdirectory(follow_actor)
 add_subdirectory(hydrodynamics)
 add_subdirectory(imu)
+add_subdirectory(force_torque)
```
Comment (chapulina) [inline src/systems/CMakeLists.txt:104]: alphabetize

File: src/systems/force_torque/ForceTorque.cc [hunk 2]
```diff
@@ -0,0 +1,347 @@
...
+
+#include "ignition/gazebo/components/ChildLinkName.hh"
+#include "ignition/gazebo/components/ForceTorque.hh"
+#include "ignition/gazebo/components/Joint.hh"
+#include "ignition/gazebo/components/JointTransmittedWrench.hh"
+#include "ignition/gazebo/components/Link.hh"
+#include "ignition/gazebo/components/Name.hh"
+#include "ignition/gazebo/components/ParentLinkName.hh"
+#include "ignition/gazebo/components/Pose.hh"
+#include "ignition/gazebo/components/ParentEntity.hh"
```
Comment (chapulina) [inline src/systems/force_torque/ForceTorque.cc:43]: alphabetize

File: src/systems/force_torque/ForceTorque.cc [hunk 3]
```diff
@@ -0,0 +1,347 @@
...
+          // X_WJ: Pose of joint in world
+          // X_WP: Pose of parent link in world
+          // X_WC: Pose of child link in world
+          // X_WS: Pose of sensor in world
+          //
+          const auto X_WP =
+              worldPose(jointLinkIt->second.jointParentLink, _ecm);
+          const auto X_WC = worldPose(jointLinkIt->second.jointChildLink, _ecm);
+          // There appears to be a bug worldPose for computing poses of //joint
+          // and its children, so we do it manually here.
```
Comment (chapulina) [inline src/systems/force_torque/ForceTorque.cc:283]: It would be good to ticket an issue for this. I believe it's because the joint is a child of a model, but its pose is expressed in the child link's frame.

File: src/systems/force_torque/ForceTorque.cc [hunk 4]
```diff
@@ -0,0 +1,347 @@
...
+          const auto X_CJ =
+              _ecm.Component<components::Pose>(jointLinkIt->second.joint)
+                  ->Data();
+          auto X_WJ = X_WC * X_CJ;
+
+          auto X_JS = _ecm.Component<components::Pose>(_entity)->Data();
+          auto X_WS = X_WJ * X_JS;
+          auto X_SP = X_WS.Inverse() * X_WP;
+          auto X_SC = X_WS.Inverse() * X_WC;
```
Comment (chapulina) [inline src/systems/force_torque/ForceTorque.cc:292]: These could be documented above too, I believe it's

```
          // X_SP: Pose of parent link in sensor frame
          // X_SC: Pose of child link in sensor frame
```

File: src/systems/force_torque/ForceTorque.cc [hunk 5]
```diff
@@ -0,0 +1,347 @@
...
+
+          math::Vector3d torque =
+              X_JS.Rot().Inverse() *
+                  msgs::Convert(jointWrench->Data().torque()) +
+              X_JS.Pos().Cross(force);
+
+          it->second->SetForce(force);
+          it->second->SetTorque(torque);
+          it->second->SetRotationParentInSensor(X_SP.Rot());
+          it->second->SetRotationChildInSensor(X_SC.Rot());
```
Comment (chapulina) [inline src/systems/force_torque/ForceTorque.cc:307]: Is it possible for the sensor pose to change w.r.t. the child link's pose?

File: test/integration/force_torque_system.cc [hunk 6]
```diff
@@ -0,0 +1,171 @@
...
+
+    const auto &wrench = wrenches.back();
+
+    const math::Vector3 expectedForce =
+        math::Vector3d{0, 0, kGravity * (kSensorMass + kWeightMass)};
+    EXPECT_EQ(expectedForce, msgs::Convert(wrench.force()));
+    EXPECT_EQ(math::Vector3d::Zero, msgs::Convert(wrench.torque()));
+    wrenches.clear();
+  }
+}
```
Comment (chapulina) [inline test/integration/force_torque_system.cc:171]: It would be good to add a test that has non-zero torque.

**Review by chapulina — APPROVED** [review]
I just have some minor comments, but in general it LGTM.

File: src/systems/force_torque/ForceTorque.cc (hunk 2 above)
Comment (mjcarroll) [reply src/systems/force_torque/ForceTorque.cc]: 54793e92

File: src/systems/force_torque/ForceTorque.cc (hunk 4 above)
Comment (mjcarroll) [reply src/systems/force_torque/ForceTorque.cc:293]: 54793e92

File: test/integration/force_torque_system.cc (hunk 6 above)
Comment (mjcarroll) [reply test/integration/force_torque_system.cc:171]: I believe the test after checks for that:

```    
    const double kMomentArm = 0.1;
    const auto &wrench = wrenches.back();
    const math::Vector3 expectedForce =
        math::Vector3d{0, 0, kGravity * (kSensorMass + kWeightMass)};
    EXPECT_NEAR(-kMomentArm * expectedForce.Z(), wrench.torque().y(), 1e-3);
```

File: src/systems/CMakeLists.txt (hunk 1 above)
Comment (mjcarroll) [reply src/systems/CMakeLists.txt]: 4c79e65d

File: test/integration/force_torque_system.cc (hunk 6 above)
Comment (azeey) [reply test/integration/force_torque_system.cc:171]: The section listening to `/force_torque2` has non-zero torque.

Comment (azeey) [reply test/integration/force_torque_system.cc:171]: Thought, I found a sign error when looking at it again. https://github.com/ignitionrobotics/ign-gazebo/pull/977/commits/416449995f450e56b3b12ab9d50a0c0eb298c993

File: src/systems/force_torque/ForceTorque.cc (hunk 5 above)
Comment (azeey) [reply src/systems/force_torque/ForceTorque.cc]: Good point! I don't it can, so I set the rotation during the creation of the sensor instead of recomputing it here every time. https://github.com/ignitionrobotics/ign-gazebo/pull/977/commits/416449995f450e56b3b12ab9d50a0c0eb298c993

File: src/systems/force_torque/ForceTorque.cc (hunk 3 above)
Comment (azeey) [reply src/systems/force_torque/ForceTorque.cc:291]: Yup, I think that's the problem. https://github.com/ignitionrobotics/ign-gazebo/issues/1042

**Comment by azeey** [comment]
CMake is finding `ign-sensors 6.0.0~pre1`, but the force-torque component doesn't seem to be there

```
  
 -- Looking for ignition-sensors6 -- found version 6.0.0~pre1
 -- Searching for dependencies of ignition-sensors6
...
 -- Looking for ignition-sensors6 - not found
    By not providing "Findignition-sensors6-force_torque.cmake" in
  
    CMAKE_MODULE_PATH this project has asked CMake to find a package
    configuration file provided by "ignition-sensors6-force_torque", but CMake
    did not find one.
  
    Could not find a package configuration file provided by
    "ignition-sensors6-force_torque" (requested version 6.0.0) with any of the
    following names:
  
      ignition-sensors6-force_torqueConfig.cmake
      ignition-sensors6-force_torque-config.cmake
```

File: test/integration/force_torque_system.cc (hunk 6 above)
Comment (chapulina) [reply test/integration/force_torque_system.cc:171]: Ahh I had missed it, sorry for the noise
